xls/modules/rle/rle.x

```diff
@@ -0,0 +1,253 @@
+// Copyright 2023 The XLS Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// This file implements a parametric RLE encoder
+//
+// The encoder uses Run Length Encoding (RLE) to compress the input stream
+// with symbols (data to compress). The output contains packets of data with
+// the symbols and the number of their occurrences in the input stream.
+
+import std
+
+// structure containing output data from an RLE encoder
+struct EncOutData<SYMB_WIDTH: u32, CNT_WIDTH: u32> {
+    symb: bits[SYMB_WIDTH], // symbol
+    cnt: bits[CNT_WIDTH],   // symbol counter
+}
+
+// structure to preserve the state of an RLE encoder for the next iteration
+pub struct RLEEncState<SYMB_WIDTH: u32, CNT_WIDTH: u32> {
+    prev: EncOutData<SYMB_WIDTH, CNT_WIDTH>, // previous symbol and symbol counter
+    flush: bool,                             // if flush was requested
+}
+
+// RLE encoder implementation
+pub proc RLEEnc<SYMB_WIDTH: u32, CNT_WIDTH: u32> {
+    symb_r: chan<bits[SYMB_WIDTH]> in;                   // input with symbols
+    flush_r: chan<bool> in;                              // input with flush request
+    data_s: chan<EncOutData<SYMB_WIDTH, CNT_WIDTH>> out; // output with data (symbol + counter)
+
+    init {(
+        RLEEncState<SYMB_WIDTH, CNT_WIDTH> {
+            prev: EncOutData<SYMB_WIDTH, CNT_WIDTH> {
+                symb: bits[SYMB_WIDTH]:0,
+                cnt: bits[CNT_WIDTH]:0,
+            },
+            flush: false,
+        }
+    )}
+
+    config (
+        symb_r: chan<bits[SYMB_WIDTH]> in,
+        flush_r: chan<bool> in,
+        data_s: chan<EncOutData<SYMB_WIDTH, CNT_WIDTH>> out,
+    ) {(symb_r, flush_r, data_s)}
+
+    next (tok: token, state: RLEEncState<SYMB_WIDTH, CNT_WIDTH>) {
+        let (flush_tok, flush) = recv_if(tok, flush_r, !state.flush, false);
+        let (data_tok, symb) = recv_if(tok, symb_r, !state.flush, bits[SYMB_WIDTH]:0);
+        let _ = join(flush_tok, data_tok);
+
+        let prev_symb_valid = state.prev.cnt != bits[CNT_WIDTH]:0;
+        let symb_differ = prev_symb_valid && (symb != state.prev.symb);
+        let overflow = state.prev.cnt == std::unsigned_max_value<CNT_WIDTH>();
+
+        let (do_send, next_state) = if (state.flush) {
```

Comment:
I wonder if this would be more readable to separate the state update logic from the communication one?

Reply:
For me, it was more obvious to think about it in this way, for example:

* if the `flush` was requested we want to send the data and our next state is ABC
* if we have an overflow or change of the symbol, this means that we have to send the data and our next state is XYZ

But of course, I can separate that if you want

Reply:
fixed